docs/FlowClassifyHook.md

```diff
@@ -0,0 +1,253 @@
+   - **Notes:** This avoids unnecessary stream processing for flows that do not require data inspection at the stream layer.
+
+2. **stream_flow_classify**
+   - **Invocation:** Called for each TCP segment on flows that were marked as needing more data by `new_flow_classify`.
+   - **Purpose:** Allows the eBPF program to inspect stream data and allow/block the connection, or request more data if classification is not yet possible.
+   - **Return values:**
+     - `FLOW_CLASSIFY_ALLOW`: The flow is permitted; no further stream inspection for this flow.
+     - `FLOW_CLASSIFY_BLOCK`: The flow is blocked; all further data is dropped (using WFP ABSORB), and the connection will be terminated or time out.
+     - `FLOW_CLASSIFY_NEED_MORE_DATA`: The current segment is allowed, but the program will be invoked again for subsequent segments until a final decision is made.
+   - **Notes:** This is only invoked for each flow segment until allow/block are returned.
+
+3. **flow_deleted**
+   - **Invocation:** Called when a flow is deleted (e.g., connection teardown) and the last action for the flow was `FLOW_CLASSIFY_NEED_MORE_DATA` (i.e., the program never returned a final allow/block decision).
+   - **Purpose:** Allows the eBPF program to perform cleanup of any per-flow state (e.g., in maps) that was maintained during inspection.
+   - **Return value:**
+     - Ignored
+   - **Notes:** This hook is only invoked for flows that were still being actively classified and not for those that were allowed immediately.
+
+```c
+/*
+ * @brief Select flows for stream layer classification
+ *
+ * Program type: \ref EBPF_PROGRAM_TYPE_FLOW_CLASSIFY
+ *
+ * Attach type(s):
+ * \ref EBPF_ATTACH_TYPE_NEW_FLOW_CLASSIFY
+ *
+ * @param[in] context \ref bpf_flow_classify_t
+ * @return ALLOW to ignore(allow) this flow, NEED_MORE_DATA for stream layer classification.
+ */
+typedef flow_classify_action_t
+new_flow_classify_hook_t(bpf_flow_classify_t* context);
+
+/*
+ * @brief Handle flow classification (stream inspection).
+ *
+ * Program type: \ref EBPF_PROGRAM_TYPE_FLOW_CLASSIFY
+ *
+ * Attach type(s):
+ * \ref EBPF_ATTACH_TYPE_STREAM_FLOW_CLASSIFY
+ *
+ * @param[in] context \ref bpf_flow_classify_t
+ * @return classification decision (allow, block, or need more data to decide).
+ */
+typedef flow_classify_action_t
+stream_flow_classify_hook_t(bpf_flow_classify_t* context);
+
+/*
+ * @brief Handle deletion of a flow that is still being classified
+ *
+ * Called for flows where the last return value was NEED_MORE_DATA
+ *
+ * Program type: \ref EBPF_PROGRAM_TYPE_FLOW_CLASSIFY
+ *
+ * Attach type(s):
+ * \ref EBPF_ATTACH_TYPE_FLOW_DELETED
+ *
+ * @param[in] context \ref bpf_flow_classify_t
+ * @return classification decision (allow, block, or need more data to decide).
+ */
+typedef flow_classify_action_t
+flow_deleted_hook_t(bpf_flow_classify_t* context);
+```
+## Architecture
```

Comment:
Consider adding more explicit details about what the hook is doing at the WFP layer. A few questions that I have:
- I see you have proposed a BPF need_more_data value, but does that correspond to the FWPS_STREAM_ACTION_NEED_MORE_DATA  value?
- The stream inspection callout has references to things like countBytesEnforced - how are we setting this value in the hook?
- Are we doing inline stream inspection only?
- We must ensure that we are following the waterfall model in our API usage. It may be good to consider adding some details about how we are doing it (as this is the crux of a stream WFP callout, and places that can cause kernel crashes if done incorrectly).